src/Context.php

```diff
@@ -45,11 +45,25 @@ public function __debugInfo()
         ];
     }
 
-    /** @param array<(int|string), mixed> $data */
-    public function withData(array $data, bool $keepExisting = true): self
+    /**
+     * @param array<(int|string), mixed> $data
+     *
+     * @throws \Exception
+     */
+    public function withData(array $data, bool $keepExisting = true, bool $recursive = true): self
     {
+        if ($keepExisting) {
+            if ($recursive) {
+                /* @var array<(int|string), mixed> */
+                $data = $this->arrayMergeRecursiveDistinct($this->data, $data);
```

Comment:
Why don't you use https://www.php.net/manual/en/function.array-merge-recursive.php?

Reply:
I was originally going for `array_merge_recursive`, but in the end, it doesn't fit the requirement. Because we only want to merge array

For example, the name becomes an array, and that is not expected :
```php
[
    'name' => [
        'my_default',
        'production',
    ]
]
```